File: mobile/android/app/src/main/java/top/onethinker/yunjian/MainViewModel.kt
package top.onethinker.yunjian

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import androidx.core.content.ContextCompat
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/** 首启物化的三态。 */
sealed interface CorpusState {
    data object Idle : CorpusState
    data class Working(val stage: MaterializationStage) : CorpusState
    data class Ready(val facts: String) : CorpusState
    data class Failed(val reason: String) : CorpusState
}

/** 语音一轮的可观测状态。 */
sealed interface VoiceState {
    data object Idle : VoiceState
    data class Listening(val detail: String) : VoiceState
    data class Finished(val detail: String) : VoiceState

    /**
     * 语音不可用，已降级到打字。
     *
     * `reason` 必须是**具体原因**而不是「语音不可用」：真机上这条断言检的正是
     * 「显示具体原因」，一句笼统的话与不显示等价。
     */
    data class Degraded(val reason: String) : VoiceState
}

data class UiState(
    val corpus: CorpusState = CorpusState.Idle,
    val query: String = "",
    val directId: String = "",
    val hits: List<SearchHit> = emptyList(),
    val searched: Boolean = false,
    val reading: PoemReading? = null,
    val reciteSession: ReciteSession? = null,
    val reciteAnswer: String = "",
    val reciteScore: ReciteScore? = null,
    val voice: VoiceState = VoiceState.Idle,
    val error: String? = null,
)

class MainViewModel(private val repository: YunjianRepository) : ViewModel() {
    private val _state = MutableStateFlow(UiState())
    val state: StateFlow<UiState> = _state.asStateFlow()

    /**
     * 首启物化。
     *
     * 语料已在本地时**也走这条路径**：`AssetResolver` 的 `already_present` 分支会立刻
     * 报出来，界面因此总有话说。跳过它会让「第二次启动界面一片空白」成为一个正常状态。
     */
    fun materialize() {
        // 三态都要挡：`Working` 是正在跑，`Ready` 是已经跑完。只挡 `Working` 时
        // Activity 重建会在已就绪之后再跑一次，而那一次与仍持有句柄的门面撞成
        // `database is locked`（真机实测）。
        if (_state.value.corpus is CorpusState.Working || _state.value.corpus is CorpusState.Ready) {
            return
        }
        _state.update { it.copy(corpus = CorpusState.Working(MaterializationStage("starting", "正在联系发布地址", null))) }
        viewModelScope.launch(Dispatchers.IO) {
            runCatching {
                repository.materialize(
                    onStage = { stage ->
                        _state.update { current ->
                            if (stage.stage == "summary") {
                                current.copy(corpus = CorpusState.Ready(stage.detail))
                            } else {
                                current.copy(corpus = CorpusState.Working(stage))
                            }
                        }
                    },
                    onDone = { failure ->
                        _state.update { current ->
                            when {
                                failure != null -> current.copy(corpus = CorpusState.Failed(failure))
                                current.corpus is CorpusState.Ready -> current
                                else -> current.copy(corpus = CorpusState.Ready(readFacts()))
                            }
                        }
                    },
                )
            }.onFailure { error ->
                // 带上异常类名：`message` 可能为 null（`UnsatisfiedLinkError` 常见），
                // 那时只写一句「语料物化失败」等于把真因丢掉。
                _state.update {
                    it.copy(
                        corpus = CorpusState.Failed(
                            error.message ?: "${error.javaClass.name}（无 message）",
                        ),
                    )
                }
            }
        }
    }

    private fun readFacts(): String =
        runCatching { repository.corpusStatusJson() }.getOrElse { "语料状态不可读：${it.message}" }

    fun onQueryChange(value: String) {
        _state.update { it.copy(query = value) }
    }

    fun onDirectIdChange(value: String) {
        _state.update { it.copy(directId = value) }
    }

    fun search() {
        val query = _state.value.query.trim()
        if (query.isEmpty()) return
        viewModelScope.launch {
            val result = withContext(Dispatchers.IO) { runCatching { repository.searchText(query) } }
            result
                .onSuccess { hits -> _state.update { it.copy(hits = hits, searched = true, error = null) } }
                .onFailure { error -> _state.update { it.copy(error = error.message, searched = true) } }
        }
    }

    fun openReading(poemId: String) {
        viewModelScope.launch {
            val result = withContext(Dispatchers.IO) { runCatching { repository.reading(poemId) } }
            result
                .onSuccess { reading -> _state.update { it.copy(reading = reading, error = null) } }
                .onFailure { error -> _state.update { it.copy(error = error.message) } }
        }
    }

    fun startRecite(poemId: String) {
        viewModelScope.launch {
            val result = withContext(Dispatchers.IO) { runCatching { repository.reciteStart(poemId) } }
            result
                .onSuccess { session ->
                    _state.update {
                        // 预填正文：「向已有内容的字段输入」这条断言要求字段一开始就不为空。
                        // 空字段测不出「输入法在已有文本上追加」这件事。
                        it.copy(reciteSession = session, reciteAnswer = "明月", reciteScore = null, error = null)
                    }
                }
                .onFailure { error -> _state.update { it.copy(error = error.message) } }
        }
    }

    fun onReciteAnswerChange(value: String) {
        _state.update { it.copy(reciteAnswer = value) }
    }

    fun submitRecite(grade: String = "good") {
        val session = _state.value.reciteSession ?: return
        val answer = _state.value.reciteAnswer
        viewModelScope.launch {
            val result = withContext(Dispatchers.IO) {
                runCatching { repository.reciteSubmit(session.poemId, answer, grade) }
            }
            result
                .onSuccess { score -> _state.update { it.copy(reciteScore = score, error = null) } }
                .onFailure { error -> _state.update { it.copy(error = error.message) } }
        }
    }

    /**
     * 一轮语音跟读。
     *
     * 三条降级路径都写成 [VoiceState.Degraded] 并带具体原因：
     *
     * 1. 未授予 `RECORD_AUDIO`；
     * 2. 本次原生库未启用 `native-voice`；
     * 3. ASR 权重目录不在（模型按需下载，未下载时不是缺陷）。
     *
     * **刻意不做的事**：不把识别结果送进评分。2026-08-11 裁决按 1800 句实测 CER 77.01%
     * 定下 `guided_practice`——只报「是否开口／停顿／相对节奏」，FSRS 等级由用户自选。
     */
    fun startVoiceRound(context: Context, poemId: String, reference: String, modelDir: String) {
        val granted = ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO) ==
            PackageManager.PERMISSION_GRANTED
        if (!granted) {
            _state.update {
                it.copy(
                    voice = VoiceState.Degraded("未授予麦克风权限（android.permission.RECORD_AUDIO 被拒绝）；已切到打字背诵"),
                )
            }
            startRecite(poemId)
            return
        }
        if (!java.io.File(modelDir).isDirectory) {
            _state.update {
                it.copy(voice = VoiceState.Degraded("ASR 权重目录不存在：$modelDir；已切到打字背诵"))
            }
            startRecite(poemId)
            return
        }
        // `checkSelfPermission` 说已授予**不等于**真能采到音。运行时权限之外还有一层
        // appops：`android:record_audio` 被拒时 `AudioRecord` 照样能建、`startRecording`
        // 也不报错，读到的是**静音流**。只看权限的产品会在这种设备状态下静静录一段空白，
        // 用户看到「正在采集」却永远没有结果——比明确降级更糟。
        // 所以先探一小段，全零即按「采集被拒」降级并说明。
        val silent = probeSilentCapture()
        if (silent != null) {
            _state.update { it.copy(voice = VoiceState.Degraded(silent)) }
            startRecite(poemId)
            return
        }
        _state.update { it.copy(voice = VoiceState.Listening("正在采集")) }
        viewModelScope.launch(Dispatchers.IO) {
            runCatching { runVoiceRound(reference, modelDir) }
                .onSuccess { detail -> _state.update { it.copy(voice = VoiceState.Finished(detail)) } }
                .onFailure { error ->
                    _state.update {
                        it.copy(voice = VoiceState.Degraded("${error.message ?: "语音会话失败"}；已切到打字背诵"))
                    }
                    startRecite(poemId)
                }
        }
    }

    /**
     * 采集能否真的拿到非静音数据。可用时返回 `null`，否则返回可直接展示的原因。
     *
     * 判据是「这一小段里有没有任何非零采样」，不是能量阈值：静音流是**逐字节全零**，
     * 而环境噪声哪怕极轻也不会全零。用阈值反而会把安静房间里的真采集误判成被拒。
     */
    private fun probeSilentCapture(): String? {
        val minBuffer = AudioRecord.getMinBufferSize(SAMPLE_RATE, CHANNEL, ENCODING)
        if (minBuffer <= 0) {
            return "音频采集不可用：AudioRecord.getMinBufferSize 返回 $minBuffer；已切到打字背诵"
        }
        val recorder = AudioRecord(
            MediaRecorder.AudioSource.MIC,
            SAMPLE_RATE,
            CHANNEL,
            ENCODING,
            maxOf(minBuffer, FRAME_SAMPLES * 2 * 4),
        )
        try {
            if (recorder.state != AudioRecord.STATE_INITIALIZED) {
                return "音频采集不可用：AudioRecord 未初始化（state=${recorder.state}）；已切到打字背诵"
            }
            recorder.startRecording()
            if (recorder.recordingState != AudioRecord.RECORDSTATE_RECORDING) {
                return "音频采集被拒：startRecording 未进入录音状态；已切到打字背诵"
            }
            val buffer = ShortArray(FRAME_SAMPLES)
            var nonZero = 0
            var read = 0
            repeat(SILENCE_PROBE_FRAMES) {
                val n = recorder.read(buffer, 0, buffer.size)
                if (n > 0) {
                    read += n
                    nonZero += (0 until n).count { index -> buffer[index] != 0.toShort() }
                }
            }
            if (read == 0) {
                return "音频采集被拒：读不到任何采样（RECORD_AUDIO 的 appops 可能为 deny）；已切到打字背诵"
            }
            if (nonZero == 0) {
                return "音频采集被拒：$read 个采样全为静音，RECORD_AUDIO 权限或 appops 未真正放开；已切到打字背诵"
            }
            return null
        } finally {
            runCatching { recorder.stop() }
            recorder.release()
        }
    }

    private fun runVoiceRound(reference: String, modelDir: String): String {
        val operation = repository.startAsr(modelDir, reference, SAMPLE_RATE)
        val minBuffer = AudioRecord.getMinBufferSize(SAMPLE_RATE, CHANNEL, ENCODING)
        val recorder = AudioRecord(
            MediaRecorder.AudioSource.MIC,
            SAMPLE_RATE,
            CHANNEL,
            ENCODING,
            maxOf(minBuffer, FRAME_SAMPLES * 2 * 4),
        )
        val outcome = StringBuilder()
        try {
            recorder.startRecording()
            val buffer = ShortArray(FRAME_SAMPLES)
            var pushed = 0
            while (pushed < FRAMES_PER_ROUND) {
                val read = recorder.read(buffer, 0, buffer.size)
                if (read <= 0) break
                operation.pushPcm(FloatArray(read) { index -> buffer[index] / 32768f }.toList())
                pushed += 1
            }
            operation.finishInput()
            // 持续拉取直到唯一终态。一次轮询超时（`null`）不是终态——把它当终态会让
            // 语音一轮在识别还没吐出 outcome 时就被判成结束。
            while (true) {
                val raw = operation.nextEvent(POLL_MS.toULong()) ?: continue
                val event = org.json.JSONObject(raw)
                when (event.optString("type")) {
                    "item" -> outcome.append(describeItem(event.getJSONObject("payload")))
                    "done" -> return outcome.ifEmptyDefault()
                    "failed" -> throw IllegalStateException(event.optString("message"))
                    "cancelled" -> throw IllegalStateException("语音会话已取消")
                }
            }
        } finally {
            runCatching { recorder.stop() }
            recorder.release()
            operation.shutdown()
        }
    }

    private fun describeItem(payload: org.json.JSONObject): String =
        when (payload.optString("type")) {
            "outcome" ->
                "开口=${payload.optBoolean("spoke")} 停顿=${payload.optInt("pause_count")} " +
                    "时长=${payload.optLong("total_ms")}ms 单路RTF=${payload.optDouble("single_rtf")}"
            else -> ""
        }

    private fun StringBuilder.ifEmptyDefault(): String =
        if (isBlank()) "识别结束但未产出 outcome" else toString()

    fun dismissError() {
        _state.update { it.copy(error = null) }
    }

    private companion object {
        const val SAMPLE_RATE = 16_000
        const val CHANNEL = AudioFormat.CHANNEL_IN_MONO
        const val ENCODING = AudioFormat.ENCODING_PCM_16BIT
        const val FRAME_SAMPLES = 1_600
        const val FRAMES_PER_ROUND = 30
        const val POLL_MS = 200L

        /** 静音探测读几帧。3 帧 × 1600 采样 ≈ 300 ms，足够区分全零流与真采集。 */
        const val SILENCE_PROBE_FRAMES = 3
    }
}

File: mobile/ios/Yunjian/MainViewModel.swift
import Foundation
import YunjianMobile

/// 首启物化的四态。与 Android 的 `CorpusState` 一一对应。
enum CorpusState: Equatable {
    case idle
    case working(MaterializationStage)
    case ready(facts: String)
    case failed(reason: String)
}

/// 语音一轮的可观测状态。与 Android 的 `VoiceState` 一一对应。
enum VoiceState: Equatable {
    case idle
    case listening(detail: String)
    case finished(detail: String)
    /// 语音不可用，已降级到打字。
    ///
    /// `reason` 必须是**具体原因**而不是「语音不可用」：验收判据检的正是
    /// `reason_names_capture_denial`，一句笼统的话与不显示等价。
    case degraded(reason: String)
}

/// 界面状态。字段与 Android 的 `UiState` 一一对应，顺序也一样——两侧的断言读同一批东西。
struct UiState: Equatable {
    var corpus: CorpusState = .idle
    var query: String = ""
    var directId: String = ""
    var hits: [SearchHit] = []
    var searched: Bool = false
    var reading: PoemReading?
    var reciteSession: ReciteSession?
    var reciteAnswer: String = ""
    var reciteScore: ReciteScore?
    var voice: VoiceState = .idle
    var error: String?
}

/// 全部界面行为。Android 是 `MainViewModel`，这里是同名同职责的 `ObservableObject`。
///
/// # 为什么状态集中在一个 `@Published`
///
/// Android 侧用单个 `MutableStateFlow<UiState>`。拆成多个 `@Published` 会让「检索成功时要
/// 同时清掉上一首阅读页」这类**成组**变更出现中间帧，而那个中间帧正是真机上逮到的缺陷形态
/// （旧正文压在新结果上面，想点的按钮跑到屏幕外，点击静默落空）。
///
/// # 尚未由 Xcode 编译验证
///
/// 本文件没有经过 Swift 编译器与真机运行（本机无 macOS）。见 `mobile/ios/README.md`。
@MainActor
final class MainViewModel: ObservableObject {
    @Published private(set) var state = UiState()

    private let repository: YunjianRepository

    /// 流式识别模型目录名。与 `models/cache/` 下的上游发布包同名，与 Android 侧相同。
    static let streamingAsrModel = "sherpa-onnx-streaming-zipformer-bilingual-zh-en-2023-02-20"

    init(repository: YunjianRepository = .shared) {
        self.repository = repository
    }

    /// 权重目录。与 Android 的 `MainActivity` 同一口径：应用容器内的模型根 + 模型名。
    var modelDir: String {
        repository.paths.modelRoot.appendingPathComponent(Self.streamingAsrModel).path
    }

    /// 首启物化。
    ///
    /// 语料已在本地时**也走这条路径**：`AssetResolver` 的 `already_present` 分支会立刻报出来，
    /// 界面因此总有话说。跳过它会让「第二次启动界面一片空白」成为一个正常状态。
    ///
    /// 三态都要挡：`working` 是正在跑，`ready` 是已经跑完。只挡 `working` 时场景重建会在
    /// 已就绪之后再跑一次，而那一次与仍持有句柄的门面撞成 `database is locked`
    /// （Android 真机实测）。
    func materialize() {
        if case .working = state.corpus { return }
        if case .ready = state.corpus { return }
        state.corpus = .working(MaterializationStage(stage: "starting", detail: "正在联系发布地址", fraction: nil))
        repository.materialize(
            onStage: { [weak self] stage in
                Task { @MainActor [weak self] in
                    guard let self else { return }
                    if stage.stage == "summary" {
                        self.state.corpus = .ready(facts: stage.detail)
                    } else {
                        self.state.corpus = .working(stage)
                    }
                }
            },
            onDone: { [weak self] failure in
                Task { @MainActor [weak self] in
                    guard let self else { return }
                    if let failure {
                        self.state.corpus = .failed(reason: failure)
                        return
                    }
                    if case .ready = self.state.corpus { return }
                    self.state.corpus = .ready(facts: self.readFacts())
                }
            }
        )
    }

    private func readFacts() -> String {
        do {
            return try repository.corpusStatusJson()
        } catch {
            return "语料状态不可读：\(YunjianRepository.describe(error))"
        }
    }

    func onQueryChange(_ value: String) { state.query = value }

    func onDirectIdChange(_ value: String) { state.directId = value }

    /// 发起一次检索。
    ///
    /// **新的检索会收掉上一首的阅读页。** 阅读页独占一屏（见 `ContentView`），不收掉它，
    /// 用户重新检索后看到的仍是上一首——Android 真机上这条曾让 `performClick` 静默落空，
    /// 并把一次装置问题记成产品 FAIL。这是产品缺陷而不是断言问题：真人遇到的是同一件事。
    func search() {
        let query = state.query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !query.isEmpty else { return }
        run {
            let hits = try self.repository.searchText(query: query)
            return { state in
                state.hits = hits
                state.searched = true
                state.error = nil
                state.reading = nil
            }
        } onFailure: { message in
            { state in
                state.error = message
                state.searched = true
            }
        }
    }

    func openReading(poemId: String) {
        let trimmed = poemId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        run {
            let reading = try self.repository.reading(poemId: trimmed)
            return { state in
                state.reading = reading
                state.error = nil
            }
        }
    }

    /// 关掉阅读页回到检索。阅读页独占一屏，这是唯一的出口。
    func closeReading() { state.reading = nil }

    func startRecite(poemId: String) {
        run {
            let session = try self.repository.reciteStart(poemId: poemId)
            return { state in
                state.reciteSession = session
                // 预填正文：「向已有内容的字段输入」这条断言要求字段一开始就不为空。
                // 空字段测不出「输入法在已有文本上追加」这件事。与 Android 同一预填值。
                state.reciteAnswer = "明月"
                state.reciteScore = nil
                state.error = nil
            }
        }
    }

    func onReciteAnswerChange(_ value: String) { state.reciteAnswer = value }

    func submitRecite(grade: String = "good") {
        guard let session = state.reciteSession else { return }
        let answer = state.reciteAnswer
        run {
            let score = try self.repository.reciteSubmit(
                poemId: session.poemId,
                answer: answer,
                grade: grade
            )
            return { state in
                state.reciteScore = score
                state.error = nil
            }
        }
    }

    /// 一轮语音跟读。
    ///
    /// 三条降级路径都写成 `.degraded` 并带具体原因，与 Android 逐条对应：
    ///
    /// 1. 未取得麦克风授权；
    /// 2. ASR 权重目录不在（模型按需下载，未下载时不是缺陷）；
    /// 3. 采集拿不到数据（授权位为真但读到静音流）。
    ///
    /// **刻意不做的事**：不把识别结果送进评分。2026-08-11 裁决按 1800 句实测 CER 77.01%
    /// 定下 `guided_practice`——只报「是否开口／停顿／相对节奏」，等级由用户自选。
    /// 判据 `auto_graded == false` 是那个裁决的可执行形态。
    func startVoiceRound(poemId: String, reference: String, modelDir: String) {
        guard VoiceCapture.isAuthorized else {
            state.voice = .degraded(reason: "未取得麦克风授权（NSMicrophoneUsageDescription 对应的权限被拒绝）；已切到打字背诵")
            startRecite(poemId: poemId)
            return
        }
        guard FileManager.default.fileExists(atPath: modelDir) else {
            state.voice = .degraded(reason: "ASR 权重目录不存在：\(modelDir)；已切到打字背诵")
            startRecite(poemId: poemId)
            return
        }
        if let silent = VoiceCapture.probeSilentCapture() {
            state.voice = .degraded(reason: silent)
            startRecite(poemId: poemId)
            return
        }
        state.voice = .listening(detail: "正在采集")
        Task.detached { [repository] in
            do {
                let detail = try Self.runVoiceRound(
                    repository: repository,
                    reference: reference,
                    modelDir: modelDir
                )
                await MainActor.run { self.state.voice = .finished(detail: detail) }
            } catch {
                let message = YunjianRepository.describe(error)
                await MainActor.run {
                    self.state.voice = .degraded(reason: "\(message)；已切到打字背诵")
                    self.startRecite(poemId: poemId)
                }
            }
        }
    }

    /// 采集 → 推送 → 拉取事件。返回可直接显示的一行测量文字。
    ///
    /// 文案与 Android 的 `describeItem` 逐字相同：`spoke=` / `pause_count=` / `total_ms=` /
    /// `single_rtf=` 四个键，验收判据从界面上读走的正是它们。改这行文案会让
    /// `voice_recitation_round_succeeds_end_to_end` 在 iOS 上永远量不到值。
    private static func runVoiceRound(
        repository: YunjianRepository,
        reference: String,
        modelDir: String
    ) throws -> String {
        let operation = try repository.startAsr(
            modelDir: modelDir,
            reference: reference,
            sampleRate: UInt32(VoiceCapture.sampleRate)
        )
        defer { operation.shutdown() }
        _ = try VoiceCapture.pushRound { frame in
            try operation.pushPcm(samples: frame)
        }
        operation.finishInput()

        var outcome = ""
        // 持续拉取直到唯一终态。一次轮询超时（`nil`）不是终态——把它当终态会让语音一轮在
        // 识别还没吐出 outcome 时就被判成结束。
        while true {
            guard let raw = operation.nextEvent(timeoutMs: 200) else { continue }
            let event = try YunjianRepository.object(raw)
            let payload = event["payload"] as? [String: Any] ?? [:]
            switch event["type"] as? String {
            case "item":
                outcome += describeItem(payload)
            case "done":
                return outcome.isEmpty ? "识别结束但未产出 outcome" : outcome
            case "failed":
                throw VoiceRoundError.failed(payload["message"] as? String ?? "语音会话失败")
            case "cancelled":
                throw VoiceRoundError.failed("语音会话已取消")
            default:
                continue
            }
        }
    }

    private static func describeItem(_ payload: [String: Any]) -> String {
        guard payload["type"] as? String == "outcome" else { return "" }
        let spoke = payload["spoke"] as? Bool ?? false
        let pauses = payload["pause_count"] as? Int ?? 0
        let totalMs = payload["total_ms"] as? Int ?? 0
        let rtf = payload["single_rtf"] as? Double ?? 0
        return "开口=\(spoke) 停顿=\(pauses) 时长=\(totalMs)ms 单路RTF=\(rtf)"
    }

    func dismissError() { state.error = nil }

    /// 把一次可能失败的原生调用搬到后台，成功与失败各自产出一个状态变更。
    ///
    /// 与 Android 的 `runCatching { withContext(Dispatchers.IO) { … } }` 同形：原生调用会打开
    /// SQLite 并可能跑数秒，放在主线程上界面会僵住，而僵住在真机上表现为「点了没反应」。
    private func run(
        _ work: @escaping () throws -> (inout UiState) -> Void,
        onFailure: @escaping (String) -> (inout UiState) -> Void = { message in { $0.error = message } }
    ) {
        Task.detached {
            let apply: (inout UiState) -> Void
            do {
                apply = try work()
            } catch {
                apply = onFailure(YunjianRepository.describe(error))
            }
            await MainActor.run { self.state.apply(apply) }
        }
    }
}

enum VoiceRoundError: Error {
    case failed(String)
}

private extension UiState {
    mutating func apply(_ change: (inout UiState) -> Void) {
        change(&self)
    }
}
